src/Functions/DateTimeTransforms.h

```diff
@@ -695,9 +696,17 @@ struct ToStartOfInterval<IntervalKind::Kind::Week>
     {
         return time_zone.toStartOfWeekInterval(time_zone.toDayNum(t), weeks);
     }
-    static UInt16 execute(Int64 t, Int64 weeks, const DateLUTImpl & time_zone, Int64 scale_multiplier)
+    static Int64 execute(Int64 t, Int64 weeks, const DateLUTImpl & time_zone, Int64 scale_multiplier, Int64 origin = 0)
     {
-        return time_zone.toStartOfWeekInterval(time_zone.toDayNum(t / scale_multiplier), weeks);
+        if (origin == 0)
+            return time_zone.toStartOfWeekInterval(time_zone.toDayNum(t / scale_multiplier), weeks);
+        else
+        {
+            if (weeks < Int64_max_value / 7) // Check if multiplication doesn't overflow Int64 value
```

Comment:
This implements an overflow check but I wonder under which conditions (SQL statement) that would trigger. `weeks` is of type `Int64` (I heard someone say is about the number of atoms in the universe), so that exception would only throw if weeks is gargantuan large.

The same argument could be made in l. 770 and l. 799.